crates/uv-requirements/src/pyproject.rs

```diff
@@ -63,71 +234,298 @@ impl Pep621Metadata {
+                        project_dir,
+                        project_sources,
+                        workspace_sources,
+                        workspace_packages,
+                    )
+                    .map_err(|err| Pep621Error::LoweringError(name, err))
+                })
+                .collect::<Result<_, Pep621Error>>()?;
+            Ok((extra_name.clone(), dependencies))
+        })
+        .collect::<Result<_, Pep621Error>>()?;
+    Ok(Requirements {
+        dependencies,
+        optional_dependencies,
+    })
+}
+
+/// Combine `project.dependencies`/`project.optional-dependencies` with `tool.uv.sources`.
+pub(crate) fn lower_requirement(
+    requirement: pep508_rs::Requirement,
+    project_name: &PackageName,
+    project_dir: &Path,
+    project_sources: &HashMap<PackageName, Source>,
+    workspace_sources: &HashMap<PackageName, Source>,
+    workspace_packages: &HashMap<PackageName, String>,
+) -> Result<Requirement, LoweringError> {
+    let source = project_sources
+        .get(&requirement.name)
+        .or(workspace_sources.get(&requirement.name))
+        .cloned();
+    if !matches!(
+        source,
+        Some(Source::Workspace {
+            // By using toml, we technically support `workspace = false`.
+            workspace: true,
+            ..
+        })
+    ) && workspace_packages.contains_key(&requirement.name)
+    {
+        return Err(LoweringError::UndeclaredWorkspacePackage);
+    }
+
+    let Some(source) = source else {
+        // Support recursive editable inclusions. TODO(konsti): This is a workspace feature.
+        return if requirement.version_or_url.is_none() && &requirement.name != project_name {
+            Err(LoweringError::UnconstrainedVersion)
+        } else {
+            Ok(Requirement::from_pep508(requirement).map_err(Box::new)?)
+        };
+    };
+
+    let source = match source {
+        Source::Git {
+            git,
+            subdirectory,
+            rev,
+            tag,
+            branch,
+        } => {
+            if matches!(requirement.version_or_url, Some(VersionOrUrl::Url(_))) {
+                return Err(LoweringError::ConflictingUrls);
+            }
+            // TODO(konsti): We know better than this enum
+            let reference = match (rev, tag, branch) {
+                (None, None, None) => GitReference::DefaultBranch,
+                (Some(rev), None, None) => {
+                    if rev.len() == 40 {
+                        GitReference::FullCommit(rev)
+                    } else {
+                        GitReference::BranchOrTagOrCommit(rev)
+                    }
+                }
+                (None, Some(tag), None) => GitReference::BranchOrTag(tag),
+                (None, None, Some(branch)) => GitReference::BranchOrTag(branch),
+                _ => return Err(LoweringError::MoreThanOneGitRef),
+            };
+
+            let mut url = Url::parse(&format!("git+{git}"))?;
```

Comment:
The docs on `RequirementSource::Git` say it's the URL _without_ the `git+` prefix. Is this correct?

Reply:
Fixed here: https://github.com/astral-sh/uv/pull/3365

Reply:
Thanks